### ISSUE TITLE
check if a token is valid

### ISSUE THREAD

**Issue by rakeen** [auth0/angular2-jwt#53]
How can I check if a token is valid or not?
`tokenNotExpired()` only checks the expiry date of the valid token. What if the token itself is not properly formatted(i.e `asd.asd.asd`)?

**Comment by escardin** [comment]
tl;dr we do everything that's feasible.

Well, if it's random garbage, it won't base64 decode. Otherwise you'd have to check the signature and we don't get that far. To be honest, it's not really something you can do client side, it's the servers job to check and potentially reject a token.

If you really want to be sure a token is valid, you'd need to validate the signature.

A signature is done in one of two ways, shared key and public key. If it's shared key, then the only way to validate a signature is to have the key used to make the signature (and you'd need some HMAC SHA libraries as well). This is a REALLY REALLY bad idea. If the client has your secret key, forget checking if a token is valid, they can make their own valid tokens!

If your token is signed via public key auth, then you can fetch the public certificate (provided it's available publicly, which imo it shouldn't be in this case) and use some libraries to validate the token. This is something that is pretty finicky to get right in my experience, and would bloat the library with dependencies for crypto. Additionally, it would be out of scope for us to fetch the cert, since we can't be sure of the url or how it'd be delivered.

Some server side projects will check that the token is in a database to see if it's been issued by them, or hasn't been blacklisted, which means a token that's technically valid, is still not usable.

The token you got from the server, assuming it decodes properly and hasn't expired, is good enough trust for the client code. If you're worried about your server issuing invalid tokens, then you have bigger problems. If you're trying to ensure that clients can't see a part of your site because of a bad token or one that doesn't contain the required permissions, well, they can just bypass your auth check, since it's just JS that they can modify at will.

Basically, other than for the convenience of the user, you can trust the token you get client side, and it should always be fully validated server side. If you protect the server side properly, then a malicious user can't get access to anything you've protected by the token (i.e. data).

**Comment by rakeen** [comment]
Thanks for taking the time to explain!
